### PR TITLE
[Flang][OpenMP] Lowering Order clause to MLIR

### PR DESCRIPTION
**Pull request by harishch4**

### CODE REVIEW

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 1]
```diff
@@ -394,6 +394,29 @@ bool ClauseProcessor::processNumThreads(
...
+bool ClauseProcessor::processOrder(mlir::omp::OrderClauseOps &result) const {
+  if (auto *clause = findUniqueClause<omp::clause::Order>()) {
+    fir::FirOpBuilder &firOpBuilder = converter.getFirOpBuilder();
+    result.orderAttr = mlir::omp::ClauseOrderKindAttr::get(
+        firOpBuilder.getContext(), mlir::omp::ClauseOrderKind::Concurrent);
+    using Order = omp::clause::Order;
+    const auto &modifier =
+        std::get<std::optional<Order::OrderModifier>>(clause->t);
+    if (modifier &&
+        *modifier == omp::clause::Order::OrderModifier::Unconstrained) {
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:406]:
```suggestion
        *modifier == Order::OrderModifier::Unconstrained) {
```

File: flang/lib/Lower/OpenMP/ClauseProcessor.cpp [hunk 2]
```diff
@@ -394,6 +394,29 @@ bool ClauseProcessor::processNumThreads(
   return false;
 }
 
+bool ClauseProcessor::processOrder(mlir::omp::OrderClauseOps &result) const {
+  if (auto *clause = findUniqueClause<omp::clause::Order>()) {
```
Comment (skatrak) [inline flang/lib/Lower/OpenMP/ClauseProcessor.cpp:398]: Nit: If you move the `using` statement above, you can simplify this statement as well.
```suggestion
  if (auto *clause = findUniqueClause<Order>()) {
```

File: flang/test/Lower/OpenMP/order-clause.f90 [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+   do i = 1, 10
+   end do
+end subroutine do_simd_order
+
+!CHECK-LABEL:   func.func @_QPdo_simd_order_parallel() {
+subroutine do_simd_order_parallel
+   !CHECK: omp.parallel {
+   !CHECK: omp.wsloop order(reproducible:concurrent) {
+   !$omp parallel
+   !$omp do simd order(reproducible:concurrent)
```
Comment (skatrak) [inline flang/test/Lower/OpenMP/order-clause.f90:58]: It seems nothing new is being checked here. Perhaps by combining both constructs we at least check that it's applied to the proper operation?
```f90
!CHECK: omp.parallel {
!CHECK: omp.wsloop order(reproducible:concurrent) {
!$omp parallel do simd order(reproducible:concurrent)
```

**Review by skatrak — COMMENTED** [review]
Thank you Harish, I think this looks good. Just minor comments.

I believe we need to process the `order` clause for the `omp.distribute` operation too. It should be as simple as adding the relevant call to `cp.processOrder()` in `genDistributeClauses()` and adding one more test in order-clause.f90.

**Review by tblah — APPROVED** [review]
LGTM with Sergio's comments

File: flang/test/Lower/OpenMP/order-clause.f90 [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+   end do
+   !CHECK: omp.distribute order(reproducible:concurrent) {
+   !$omp teams distribute order(reproducible:concurrent)
+   do i=1,10
+   end do
+   !CHECK: omp.distribute order(unconstrained:concurrent) {
+   !$omp teams distribute order(unconstrained:concurrent)
+   do i = 1, 10
+   end do
+  end subroutine
```
Comment (skatrak) [inline flang/test/Lower/OpenMP/order-clause.f90:76]: Ultra-nit: Missing line ending.

**Review by skatrak — APPROVED** [review]
Thank you for addressing my comments, LGTM!

**Pull request merged**